transport/roundtrip_test.go

```diff
@@ -0,0 +1,152 @@
+}
+
+// handlerFunc wraps a function into a transport.Handler
+type handlerFunc func(context.Context, *transport.Request, transport.ResponseWriter) error
+
+func (f handlerFunc) Handle(ctx context.Context, r *transport.Request, w transport.ResponseWriter) error {
+	return f(ctx, r, w)
+}
+
+// httpTransport implements a roundTripTransport for HTTP.
+type httpTransport struct{ t *testing.T }
+
+func (ht httpTransport) WithHandler(h transport.Handler, f func(transport.Outbound)) {
+	i := http.NewInbound("127.0.0.1:0")
+	require.NoError(ht.t, i.Start(h), "failed to start")
+	defer i.Stop()
+
+	addr := fmt.Sprintf("http://%v/", i.Addr().String())
+	o := http.NewOutbound(addr)
+	f(o)
+}
+
+// tchannelTransport implements a roundTripTransport for TChannel.
+type tchannelTransport struct{ t *testing.T }
+
+func (tt tchannelTransport) WithHandler(h transport.Handler, f func(transport.Outbound)) {
+	serverOpts := testutils.NewOpts().SetServiceName(testService)
+	clientOpts := testutils.NewOpts().SetServiceName(testCaller)
+	testutils.WithServer(tt.t, serverOpts, func(ch *tchannel.Channel, hostPort string) {
+		i := tch.NewInbound(ch)
+		require.NoError(tt.t, i.Start(h), "failed to start")
+		defer i.Stop()
+		// ^ the server is already listening so this will just set up the
+		// handler.
+
+		client := testutils.NewClient(tt.t, clientOpts)
+		o := tch.NewOutbound(client, tch.HostPort(hostPort))
+
+		f(o)
+	})
+}
+
+func TestSimpleRoundTrip(t *testing.T) {
+	tests := []roundTripTransport{
+		httpTransport{t},
+		tchannelTransport{t},
+	}
+
+	request := &transport.Request{
+		Caller:    testCaller,
+		Service:   testService,
+		Procedure: "hello",
+		Encoding:  raw.Encoding,
+		Headers:   transport.Headers{"token": "1234"},
+		TTL:       200 * time.Millisecond, // TODO use default
+		Body:      bytes.NewReader([]byte("world")),
+	}
+	requestMatcher := transporttest.NewRequestMatcher(t, request)
+
+	// Matches the response that we send from the fake handler
+	responseMatcher := transporttest.NewResponseMatcher(t, &transport.Response{
+		Headers: transport.Headers{"status": "ok"},
+		Body:    ioutil.NopCloser(bytes.NewReader([]byte("hello, world"))),
+	})
+
+	h := handlerFunc(func(_ context.Context, r *transport.Request, w transport.ResponseWriter) error {
+		require.True(t, requestMatcher.Matches(r), "request did not match")
+
+		w.AddHeaders(transport.Headers{"status": "ok"})
+		_, err := w.Write([]byte("hello, world"))
+		require.NoError(t, err, "failed to write response")
+
+		return nil
+	})
+
+	rootCtx := context.Background()
+	for _, tt := range tests {
+		ctx, _ := context.WithTimeout(rootCtx, time.Millisecond*200)
+		// TODO(abg): should be picked up from TTL if unspecified
+
+		tt.WithHandler(h, func(o transport.Outbound) {
+			// We need to restore the body into a readable io.Reader for the
+			// next iteration.
+			var body bytes.Buffer
+			request.Body = io.TeeReader(request.Body, &body)
```

Comment:
from previous PR, have a `getRequest` function that returns a request rather than trying to unread the body